completed homework from 01.09

DTB4 opened the pull request:
- two new repositories
- add new middle format models for product and supplier
- complete daemon process that receive time interval and renew suppliers and products every "var n int" seconds
- parser daemon search DB for existed sup`s ad prod`s and can either create or edit existed entity
- daemon works in goroutine and call new goroutine for every suppliers and every product needed to be created or edited

DTB4 committed:
completed homework from 01.09
- two new repositories
- add new middle format models for product and supplier
- complete daemon process that receive time interval and renew suppliers and products every "var n int" seconds
- parser daemon search DB for existed sup`s ad prod`s and can either create or edit existed entity
- daemon works in goroutine and call new goroutine for every suppliers and every product needed to be created or edited

DTB4 committed:
fixed problem with too much idle connections to DB
closed Rows() in all .query functions in repos
add transactions for every .exec function in repos
fixed errors catching in repositories products and suppliers

pechenini commented on services/menu_parser_daemon.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func (m MenuParser) TimedParsing(frequencySeconds int) {
+	for i := 0; ; i++ {
```
можно просто оставить for

DTB4 replied:
временный реликт для подсчета циклов (принято, удалено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func (p ProductsRepository) EditProduct(product *models.Product) (sql.Result, error) {
+	tx, err := p.db.Begin()
```
тебе не нужно оборачивать 1 update в транзакцию

DTB4 replied:
понял, транзакции для комплексных, множественных операций (принято, удалено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func (p ProductsRepository) DeleteProduct(id int) (sql.Result, error) {
+	tx, err := p.db.Begin()
```
не нужно оборачивать одну операцию в транзакцию

DTB4 replied:
понял, транзакции для комплексных, множественных операций (принято, удалено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func (p ProductsRepository) DeleteAllProducts() (sql.Result, error) {
+	tx, err := p.db.Begin()
```
не оборачивай одну операцию в транзакцию

DTB4 replied:
понял, транзакции для комплексных, множественных операций (принято, удалено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func (p ProductsRepository) CreateNewProduct(product *models.Product) (sql.Result, error) {
+	tx, err := p.db.Begin()
```
не оборачивай одну операцию в транзакцию, эта ошибка у тебя во всех реозиториях

DTB4 replied:
понял, транзакции для комплексных, множественных операций (принято, удалено)

pechenini commented on services/menu_parser_daemon.go:
```diff
@@ -0,0 +1,188 @@
+package services
```
выдели отдельный пакет под парсинг, ну нужно городить кучу всего внутри services, надо разбивать по логическим темам

DTB4 replied:
это да, мало ли что еще парсить зададут
пока что сложно сходу придумывать как раскинуть какой то новый функционал по папкам и файлам так что бы это было не просто для читаемости а и грамотно структурировано (принято, думаю как лучше реализовать...)

DTB4 replied:
Выделил в отдельный пакет (принято, изменено)

pechenini commented on services/menu_parser_daemon.go:
```diff
@@ -0,0 +1,188 @@
+	TimedParsing(frequencySeconds int)
+	productWork(products *[]models.ParserProduct, restID int)
+	restaurantsWork(restaurants *[]models.ParserRestaurant)
+	//getAllRestaurants () (*[]models.ParserRestaurant, error)
```
почему закомменченый код в пре

DTB4 replied:
Сомневался делать ли это частью интерфейса и в итоге забыл
Буду переписывать и менять структуру всего пакета поэтому пока удаляю 
(принято, удалено)

pechenini commented on services/menu_parser_daemon.go:
```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+func getAllRestaurants() ([]models.ParserRestaurant, error) {
```
сделай функции, которые делают запросы на урл часть какой-то структуры по типу type FoodApi struct{}

DTB4 replied:
супер
а я не мог придумать как эти методы вообще к общей структуре привязать, т.к. они вроде ничего из неё не берут
А если в неё закидывать например список УРЛов то будет логично их и использовать через структуру
(принято, в процессе реализации...)

DTB4 replied:
добавил в эти функции структуру получателя из которой они берут УРЛ и формат стринг для получения меню
УРЛ, формат стринг и задержка парсинга (в секундах) теперь подтягиваются из .env файла и кладутся в структуру парсера при его создании

pechenini commented on services/menu_parser_daemon.go:
```diff
@@ -0,0 +1,188 @@
+	return &product
+}
+
+func (m MenuParser) restaurantsWork(restaurants *[]models.ParserRestaurant) {
```
имя метода не совсем понятное

DTB4 replied:
без вопросов) (принято, поменяно)

pechenini commented on services/menu_parser_daemon.go:
```diff
@@ -0,0 +1,188 @@
+			m.logger.ErrorLog("Fail to get restaurants", err)
+		}
+		m.logger.InfoLog("iteration ", i)
+		go m.restaurantsWork(&restaurants)
```
не тут горутину нужно запускать, этот код как раз должен быть блокирующим

асинхронный запуск горутин должен быть для получения меню разных ресторанов

DTB4 replied:
вот тут я крепко налажал
Если бы не ревью, я наверное не скоро заметил что каждая горутина повторяет одну и ту же работу
Тут вызов горутины удаляю. Удобного еста для его переноса пока что нет, так что оставлю пока однопоточное выполнение. Когда буду переписывать структуру парсера выделю удобное место для старта горутин по каждому ресторану
(принято, удалено)

pechenini commented on services/menu_parser_daemon.go:
```diff
@@ -0,0 +1,188 @@
+		if err != nil {
+			m.logger.ErrorLog("Fail to get restaurants", err)
+		}
+		m.logger.InfoLog("iteration ", i)
```
бесполезный лог

DTB4 replied:
удален вместе с другими реликтовыми костылями для тестирования и дебага (принято, удалено)

pechenini commented on services/menu_parser_daemon.go:
```diff
@@ -0,0 +1,188 @@
+
+func (m MenuParser) restaurantsWork(restaurants *[]models.ParserRestaurant) {
+	for i := range *restaurants {
+		dbSupplier, err := m.restaurantRepo.SearchSupplierByID((*restaurants)[i].ID)
```
этот код должен быть запущен в отдельной горутине для каждого ресторана

DTB4 replied:
перенес цикл в вызывающую функцию и для каждого перебора ресторана запускал горутину с вызовом этой функции (принято, исправлено)

pechenini commented on services/menu_parser_daemon.go:
```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+func (m MenuParser) productWork(products *[]models.ParserProduct, restID int) {
```
имя метода ни о чем не говорит нам

DTB4 replied:
действительно (принято, поменяно)

pechenini commented on repository/order_products_repo.go:
```diff
@@ -0,0 +1,65 @@
+	db *sql.DB
+}
+
+func (op OrderProductsRepository) CreateNewOrderProduct(orderProduct *models.OrderProduct) (sql.Result, error) {
```
зачем New в методе? и так понятно, что это новая сущность в бд. оставь просто Create. И так  понятно, что ты хочешь создать OrderProduct по имени репозитория

DTB4 replied:
излишнее описание (принято, исправлено)

pechenini commented on repository/order_products_repo.go:
```diff
@@ -0,0 +1,65 @@
+	return result, nil
+}
+
+func (op OrderProductsRepository) GetOrderProductsByOrderID(id int) (*[]models.OrderProduct, error) {
```
GetByOrderID

DTB4 replied:
излишнее описание (принято, исправлено)

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/order_products_repo.go:
```diff
@@ -0,0 +1,65 @@
+	return &orderProductSlice, nil
+}
+
+func (op OrderProductsRepository) EditOrderProduct(orderProduct *models.OrderProduct) (sql.Result, error) {
```
Edit или Update

DTB4 replied:
перестарался при именовании. Сделал Update для большего соответствия (crud)
(принято, упрощено)

pechenini commented on repository/order_products_repo.go:
```diff
@@ -0,0 +1,65 @@
+	return result, nil
+}
+
+func (op OrderProductsRepository) DeleteOrderProductByIDs(orderID, productID int) (sql.Result, error) {
```
Delete

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/order_repo.go:
```diff
@@ -15,25 +15,24 @@ func NewOrderRepository(db *sql.DB) *OrderRepository {
 }
 
-func (or OrderRepository) CreateNewOrder(order *models.Order) error {
+func (or OrderRepository) CreateNewOrder(order *models.Order) (sql.Result, error) {
```
Create

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/order_repo.go:
```diff
@@ -69,40 +68,18 @@ func (or OrderRepository) GetUserOrders(userID int) (*[]models.Order, error) {
-func (or OrderRepository) GetOrderProducts(orderID int) (*[]models.OrderProduct, error) {
-	var orderProductSlice []models.OrderProduct
-	rows, err := or.db.Query("SELECT * FROM order_product WHERE order_id=?", orderID)
+func (or OrderRepository) EditOrder(order *models.Order) (sql.Result, error) {
```
Update

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/order_repo.go:
```diff
@@ -69,40 +68,18 @@ func (or OrderRepository) GetUserOrders(userID int) (*[]models.Order, error) {
 
-func (or OrderRepository) CreateOrderProducts(orderProducts *[]models.OrderProduct) error {
-	transaction, err := or.db.Begin()
+func (or OrderRepository) DeleteOrder(id int) (sql.Result, error) {
```
Delete

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+	db *sql.DB
+}
+
+func (p ProductsRepository) CreateNewProduct(product *models.Product) (sql.Result, error) {
```
Create

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+	return result, nil
+}
+
+func (p ProductsRepository) GetProductByID(id int) (*models.Product, error) {
```
GetById

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+	return &product, nil
+}
+
+func (p ProductsRepository) GetAllProducts() (*[]models.Product, error) {
```
GetAll

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+	return &products, nil
+}
+
+func (p ProductsRepository) GetAllProductsBySupplierID(id int) (*[]models.Product, error) {
```
GetBySupplierID

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+	return &products, nil
+}
+
+func (p ProductsRepository) EditProduct(product *models.Product) (sql.Result, error) {
```
Update

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+	return result, nil
+}
+
+func (p ProductsRepository) DeleteProduct(id int) (sql.Result, error) {
```
Delete

DTB4 replied:
перестарался при именовании (принято, упрощено)

pechenini commented on repository/products_repo.go:
```diff
@@ -0,0 +1,194 @@
+	return result, nil
+}
+
+func (p ProductsRepository) DeleteAllProducts() (sql.Result, error) {
```
Truncate

DTB4 replied:
Это не совсем то что я делаю этой функцией
Трункейт у меня не работает из за слегка кривых зависимостей в базе
но вообще планировался именно этот функционал так что (принято, поменяно)

pechenini replied:
А вообще тут надо делать soft delete(почитай что это такое)

pechenini commented on repository/suppliers_repo.go:
```diff
@@ -0,0 +1,162 @@
+	db *sql.DB
+}
+
+func (s SupplierRepository) CreateNewSupplier(supplier *models.Supplier) (sql.Result, error) {
```
все те же замечания, что и для остальных репозиториев

DTB4 replied:
перестарался при именовании (принято, упрощено)
для всех остальных репозиториев так же провел работу

DTB4 reviewed (commented):
no need to make transaction for single sql execution call

DTB4 commented on repository/products_repo.go:
```diff
@@ -74,52 +95,100 @@ func (p ProductsRepository) GetAllProductsBySupplierID(id int) (*[]models.Produc
+	}
 	return result, nil
 }
 
```
too much whitespace

DTB4 committed:
-fixed excessive naming of CRUD functions in repositories
-remove transactions for single sql execution calls
-delete temporary solution for counting and logging iterations

DTB4 committed:
-fixed excessive naming of CRUD functions in repositories
-fixed goroutine calls in parser daemon, now they useful
-fixed naming of parsing daemon function for better understanding their usage

DTB4 committed:
- correct rest and prod models for parsing "image" strings
- minor adds to parser and repo due to new models variables

DTB4 committed:
- make separate package for parser
- functions getallrestaurants and getProductsFromRestByID now get URL from m MenuParser structure
- parsing delay(sec), url and format string for access to menu now getting from .env file

DTB4 committed:
add softdelete and softdelete all
rewrite parser for soft delete instead of Update
